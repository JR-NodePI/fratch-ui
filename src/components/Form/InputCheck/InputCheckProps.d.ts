import { InputCheckPosition } from './InputCheckConstants';

export type InputCheckProps = {
  checked?: boolean;
  className?: string;
  disabled?: boolean;
  label: string;
  onChange?: (checked: boolean) => void;
  position?: (typeof InputCheckPosition)[keyof typeof InputCheckPosition];
};
