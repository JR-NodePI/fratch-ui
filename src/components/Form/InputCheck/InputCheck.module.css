.checkbox {
  display: inline-block;
  display: flex;
  align-items: center;
  gap: 4px;
  height: 32px;
}

.checkbox input {
  width: 20px;
  height: 20px;
}

.checkbox input {
  appearance: none;
  border: 1px solid var(--ft-form-item-border-color);
  border-radius: 3px;
  box-shadow: var(--ft-box-shadow-1), inset var(--ft-box-shadow-3);
  background-color: var(--ft-color-white);
  transition: box-shadow 0.2s ease-in-out;
}
.checkbox input:active,
.checkbox input:hover {
  box-shadow: var(--ft-box-shadow-2);
}
.checkbox input:checked {
  box-shadow: var(--ft-box-shadow-2);
}

.checkbox input:checked::before {
  content: '';
  position: relative;
  display: block;
  width: 10px;
  height: 10px;
  top: 4px;
  left: 4px;
  background-color: var(--ft-color-link);
  border-radius: 1px;
}
