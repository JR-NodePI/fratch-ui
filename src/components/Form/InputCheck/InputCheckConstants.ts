export const InputCheckPosition = Object.freeze({
  LEFT: 'left',
  RIGHT: 'right',
} as const);
