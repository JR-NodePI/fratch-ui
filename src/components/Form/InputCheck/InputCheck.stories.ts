import type { Meta, StoryObj } from '@storybook/react';

import InputCheck from './InputCheck';

const meta = {
  title: 'Example/Form/InputCheck',
  component: InputCheck,
  // tags: ['autodocs'],
  argTypes: {
    checked: {
      control: false,
      description: 'The initial checked state',
    },
  },
} satisfies Meta<typeof InputCheck>;

export default meta;
type Story = StoryObj<typeof InputCheck>;

export const _InputCheck: Story = {
  args: {
    label: 'Label input checkbox',
    checked: true,
  },
};
