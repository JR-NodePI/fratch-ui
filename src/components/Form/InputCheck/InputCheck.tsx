import { forwardRef, useEffect, useRef, useState } from 'react';

import { c } from '../../../helpers/classNameHelpers';
import { Icons } from '../..';
import { InputCheckPosition } from './InputCheckConstants';
import type { InputCheckProps } from './InputCheckProps';

import styles from './InputCheck.module.css';

const InputCheck = forwardRef<HTMLInputElement, InputCheckProps>(
  (
    {
      checked,
      className,
      disabled,
      label,
      onChange,
      position = InputCheckPosition.LEFT,
    }: InputCheckProps,
    ref: React.ForwardedRef<HTMLInputElement>
  ) => {
    const inputRef = useRef<HTMLInputElement>(null);
    const [isChecked, setIsChecked] = useState<boolean>(checked ?? false);

    if (ref == null) {
      ref = inputRef;
    }

    useEffect(() => {
      if (typeof ref === 'object' && ref?.current != null) {
        ref.current.checked = isChecked;
      }
    }, [isChecked, ref]);

    const handleOnChange = (
      event: React.ChangeEvent<HTMLInputElement>
    ): void => {
      onChange && onChange(event.target.checked);
      setIsChecked(event.target.checked);
    };

    return (
      <label className={c(styles.checkbox, styles[position], className)}>
        <input
          ref={ref}
          type="checkbox"
          disabled={disabled}
          onChange={handleOnChange}
        />
        <Icons.IconCheck className={c(styles.icon_check)} />
        <span>{label}</span>
      </label>
    );
  }
);

export default InputCheck;
