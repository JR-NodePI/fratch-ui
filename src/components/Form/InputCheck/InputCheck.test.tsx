import { createRef } from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import InputCheck from './InputCheck';
import { type InputCheckProps } from './InputCheckProps';

describe('InputCheck', () => {
  const initialProps: InputCheckProps = {
    label: 'mock label',
  };

  const getComponent = (
    props: InputCheckProps = initialProps,
    ref?: React.ForwardedRef<HTMLInputElement>
  ) => <InputCheck ref={ref} {...props} />;

  const setup = (
    props = initialProps,
    ref?: React.ForwardedRef<HTMLInputElement>
  ) => render(getComponent(props, ref));

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should be checked', async () => {
    setup({
      ...initialProps,
      checked: true,
    });
    expect(screen.getByRole('checkbox')).toBeChecked();
  });

  it('should de disabled', () => {
    setup({
      ...initialProps,
      disabled: true,
    });
    expect(screen.getByRole('checkbox')).toBeDisabled();
  });

  it('should render with className', () => {
    const { container } = setup({
      ...initialProps,
      className: 'mock-class',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with label', () => {
    const { container } = setup({
      ...initialProps,
      label: 'mock label',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render position right', () => {
    const { container } = setup({
      ...initialProps,
      position: 'right',
    });
    expect(container).toMatchSnapshot();
  });

  it('should perform onChange', async () => {
    const label = 'mock-test-label';
    const onChange = vi.fn();

    setup({ ...initialProps, label, onChange });
    await userEvent.click(screen.getByLabelText(label));

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(true);
  });

  it('should set the forwarded ref', async () => {
    const ref = createRef<HTMLInputElement>();
    setup(initialProps, ref);

    expect(ref).toMatchSnapshot();
  });
});
