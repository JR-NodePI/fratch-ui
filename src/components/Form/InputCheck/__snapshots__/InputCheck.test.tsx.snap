// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputCheck > should render position right 1`] = `
<div>
  <label
    class="_checkbox_29b3a3 _right_29b3a3"
  >
    <input
      type="checkbox"
    />
    <svg
      class="_icon_2be8fc _primary_2be8fc _icon_check_29b3a3"
      fill="none"
      viewBox="-1 -3 18 18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 5.917 5.724 10.5 15 1.5"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="3"
      />
    </svg>
    <span>
      mock label
    </span>
  </label>
</div>
`;

exports[`InputCheck > should render properly 1`] = `
<div>
  <label
    class="_checkbox_29b3a3 _left_29b3a3"
  >
    <input
      type="checkbox"
    />
    <svg
      class="_icon_2be8fc _primary_2be8fc _icon_check_29b3a3"
      fill="none"
      viewBox="-1 -3 18 18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 5.917 5.724 10.5 15 1.5"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="3"
      />
    </svg>
    <span>
      mock label
    </span>
  </label>
</div>
`;

exports[`InputCheck > should render with className 1`] = `
<div>
  <label
    class="_checkbox_29b3a3 _left_29b3a3 mock-class"
  >
    <input
      type="checkbox"
    />
    <svg
      class="_icon_2be8fc _primary_2be8fc _icon_check_29b3a3"
      fill="none"
      viewBox="-1 -3 18 18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 5.917 5.724 10.5 15 1.5"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="3"
      />
    </svg>
    <span>
      mock label
    </span>
  </label>
</div>
`;

exports[`InputCheck > should render with label 1`] = `
<div>
  <label
    class="_checkbox_29b3a3 _left_29b3a3"
  >
    <input
      type="checkbox"
    />
    <svg
      class="_icon_2be8fc _primary_2be8fc _icon_check_29b3a3"
      fill="none"
      viewBox="-1 -3 18 18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 5.917 5.724 10.5 15 1.5"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="3"
      />
    </svg>
    <span>
      mock label
    </span>
  </label>
</div>
`;

exports[`InputCheck > should set the forwarded ref 1`] = `
{
  "current": <input
    type="checkbox"
  />,
}
`;
